lib/InitModelTypes/InitModelTypes.cpp

```diff
@@ -42,53 +40,47 @@ using llvm::BasicBlock;
 using llvm::Function;
 using llvm::Instruction;
 
-using llvm::cast;
 using llvm::dyn_cast;
-using llvm::isa;
-
-using model::Binary;
-using model::QualifiedType;
 
 template<typename T>
 using RPOT = llvm::ReversePostOrderTraversal<T>;
 
-using TypeVector = llvm::SmallVector<QualifiedType, 8>;
-using ModelTypesMap = std::map<const llvm::Value *, const model::QualifiedType>;
+using TypeVector = llvm::SmallVector<model::UpcastableType, 8>;
+using ModelTypesMap = std::map<const llvm::Value *,
+                               const model::UpcastableType>;
 
-/// Map each llvm::Argument of the given llvm::Function to its
-/// QualifiedType in the model.
+/// Map each llvm::Argument of the given llvm::Function to its type in the model
 static void addArgumentsTypes(const llvm::Function &LLVMFunc,
-                              const model::TypeDefinition *Prototype,
-                              const Binary &Model,
+                              const abi::FunctionType::Layout &Layout,
+                              const model::Binary &Model,
                               ModelTypesMap &TypeMap,
                               bool PointersOnly) {
 
-  const auto Layout = abi::FunctionType::Layout::make(*Prototype);
-
-  const auto IsNonShadow = [](const abi::FunctionType::Layout::Argument &A) {
-    using namespace abi::FunctionType::ArgumentKind;
-    return A.Kind != ShadowPointerToAggregateReturnValue;
-  };
-
-  auto NumArgs = LLVMFunc.arg_size();
-  size_t NumNonShadowArgs = llvm::count_if(Layout.Arguments, IsNonShadow);
-  revng_assert(NumNonShadowArgs == NumArgs);
-  auto NonShadowArgs = llvm::make_filter_range(Layout.Arguments, IsNonShadow);
+  using AK = abi::FunctionType::ArgumentKind::Values;
+  constexpr auto SPTAR = AK::ShadowPointerToAggregateReturnValue;
+  auto NonShadowArgs = std::ranges::subrange(Layout.Arguments.begin(),
+                                             Layout.Arguments.end());
+  if (!Layout.Arguments.empty() && Layout.Arguments[0].Kind == SPTAR) {
+    revng_assert(LLVMFunc.arg_size() + 1 == Layout.Arguments.size());
+    NonShadowArgs = std::ranges::subrange(std::next(Layout.Arguments.begin()),
+                                          Layout.Arguments.end());
+  } else {
+    revng_assert(LLVMFunc.arg_size() == Layout.Arguments.size());
+  }
 
   for (const auto &[ArgModelType, LLVMArg] :
        llvm::zip_first(NonShadowArgs, LLVMFunc.args())) {
-
-    QualifiedType ArgQualifiedType = ArgModelType.Type;
-    if (not PointersOnly or ArgQualifiedType.isPointer())
-      TypeMap.insert({ &LLVMArg, std::move(ArgQualifiedType) });
+    if (not PointersOnly or ArgModelType.Type->isPointer())
+      TypeMap.insert({ &LLVMArg, ArgModelType.Type.copy() });
```

Comment:
Removing a move in favor of `.copy`. Is that you ivan?

Reply:
It's inside a loop.